### ISSUE TITLE
Out of Order read error

### ISSUE THREAD

**Issue by ryanjameslett** [jcupitt/libvips#875]
Continuation from: https://github.com/jcupitt/libvips/issues/865

Hi. I'm working through this problem slowly bug now I'm getting a new issue that I can't seem to resolve. I've been digging through the docs for quite some time now looking for how I'm messing this up but have run out of time to be messing around. 

So currently my vips.h code looks like this:

```
int
vips_embed_go(VipsImage *in, VipsImage **out, int x, int y, int width, int height, VipsArrayDouble *bg_array) {
    return vips_embed(in, out,
            x, y,
            width, height,
            "extend", VIPS_EXTEND_BACKGROUND,
            "background", bg_array,
            NULL);
}

int
vips_avg_go(VipsImage *in, double *mean) {
    // I'm passing the double in but this also fails if I create the var here 
    return vips_avg(in, mean, NULL);
}
```

The background code you helped me with works as intended. I'm creating an array of doubles to send to the embed function with hard-coded values. 

In my attempts to find an average background color I have tried a couple things but I'm getting these 'out of order read' errors for both VipsJpeg and vipspng. This error seems to be related to the incoming image since I've tried what I figured would be an easier version by looking for the average, sending in the exact same image as I'm sending to the vips_embed function. 

Note: I'm following the example at the bottom of this page: https://jcupitt.github.io/libvips/API/current/using-from-c.html 

```
func vipsExtendImage(img **C.struct__VipsImage, xsize int, ysize int, width int, height int, bg_array **C.struct__VipsArrayDouble) error {
	var out *C.struct__VipsImage
	var x_pos int
	var y_pos int
	x_pos = ((width - xsize) / 2)
	y_pos = ((height - ysize) / 2)

	var mean C.double
	C.vips_avg_go(*img, &mean)

	if C.vips_embed_go(*img, &out,
		C.int(x_pos), C.int(y_pos),
		C.int(width), C.int(height),
		*bg_array) != 0 {
		return vipsError()
	}

	C.swap_and_clear(img, out)
	return nil
}
```

The resulting errors look like this:

```
2018/02/07 03:59:08 | 500 | VipsJpeg: out of order read at line 704
VipsJpeg: out of order read at line 704
VipsJpeg: out of order read at line 704
VipsJpeg: out of order read at line 704
vips2png: unable to write to buffer
2018/02/07 03:59:08 | 500 | VipsJpeg: out of order read at line 704
VipsJpeg: out of order read at line 704
VipsJpeg: out of order read at line 704
VipsJpeg: out of order read at line 704
vips2png: unable to write to buffer
```

**Comment by jcupitt** [comment]
Hello, looks like you've hit a bug that was fixed a few weeks ago. Try 8.6.2, it should work:

https://github.com/jcupitt/libvips/releases/tag/v8.6.2

`
* fix race in vips_sink() for threaded read of sequential images
`

**Comment by ryanjameslett** [comment]
I rebuilt this using 8.6.2 and the issue is still present. I'm currently rebuilding under varying conditions to see if this issue fixes itself.

**Comment by ryanjameslett** [comment]
So I've gone and rebuilt this thing with different compilation flags and different versions (even tried 8.7.0) and no luck. Is there anything I can send include in this ticket that might help diagnose the issue?

**Comment by jcupitt** [comment]
Oh dear, I'm sorry it's not working.

I need a small (ideally) self-contained program that shows the problem. This is the test I used for the bug:

```python
import sys
import pyvips
import random

for filename in sys.argv[1:]:
    image = pyvips.Image.new_from_file(filename, access='sequential')
    x = random.randint(0, image.width - 2)
    y = random.randint(0, image.height - 2)
    w = random.randint(1, image.width - x)
    h = random.randint(1, image.height - y)
    print 'filename =', filename, 'avg =', image.crop(x, y, w, h).avg()
```

ie. open a lot of images and find the average of a lot of random areas. 

That program fails occasionally with 8.6.1 with the out of order error you had, but works fine with 8.6.2 (and 8.7.0 master) thanks to the race fix.

**Comment by jcupitt** [comment]
Ah wait, I think I understand.

Are you opening the source image in sequential access mode? If you do that, you can only loop over the image once, since it will be streamed though your system. Once you've read it, it's all used up and the pixels have vanished.

Your code is reading it twice. First you scan te whole image with `avg` to find the pixel average, then when you write the output image, you will need to loop over all the pixels again. When it tries the second loop, it attempts to fetch pixels from the top of the image, but of course the image read point is at the end after processing the `avg`, so you get the error.

I think you have two choices:

1. Open the image without the `access="sequential"` flag. This will decode the whole image into memory or a disc file and then process from that.

2. Open the image with `access="seq"`, but do it twice. Open once for the `avg` scan, then unref it, and open again for the write phase. You'll decode the image twice, but you will save a lot of memory or disc.

**Comment by ryanjameslett** [comment]
Hey, thanks again for your responsiveness on this. I'll give that a shot. If this doesn't work for whatever reason I'll probably reach out to darthsim to see if I'm doing something silly within the context of the go server. Either way I'll let you know how it goes. Thanks :+1:

**Comment by ryanjameslett** [comment]
Turns out there's a 3rd option. In the go code, there's a function that I didn't understand until now. 

```
func vipsImageCopyMemory(img **C.struct__VipsImage) error {
	var tmp *C.struct__VipsImage
	if tmp = C.vips_image_copy_memory(*img); tmp == nil {
		return vipsError()
	}
	C.swap_and_clear(img, tmp)
	return nil
}
```

I was confused as to why one would just copy data from one memory location to another and then swap them back until your last message. Looking at this function again, it is called every time a bunch of sequential image operations are performed, so I gave it a shot and called it before my stuff and it all started working.

Thanks again so much for your assistance.
